design-proposals/klip-56-schema-id-in-create-statements.md

```diff
@@ -0,0 +1,97 @@
+SELECT key, field1, field_2 FROM source_stream
+EMIT CHANGES;
+```
+```roomsql
+CREATE TABLE table_name WITH (key_schema_id=1, value_schema_id=2, key_format='avro', value_format='avro') AS 
+SELECT key, COUNT(field1) AS count_field FROM source_stream
+GROUP BY key;
+```
+
+### Schema ids in `CREATE * WITH`
+```roomsql
+CREATE STREAM stream_name WITH (kafka_topic='topic_name', key_schema_id=1, value_schema_id=2, 
+partitions=1, key_format='avro', value_format='avro');
+```
+```roomsql
+CREATE TABLE table_name WITH (kafka_topic='table_name', key_schema_id=1, value_schema_id=2, 
+partitions=1, key_format='avro', value_format='avro');
+```
+
+## Design
+### Validations needed when `*_schema_id` presents
+* `CREATE * WITH` command
+  * Corresponding `key_format` or `value_format` must exist and the format must support `SCHEMA_INFERENCE` (protobuf, avro, json_sr format currently).
+  * The fetched schema format from _schema_registry_ must match specified format in `WITH` clause. For example, if schema format for schema id in _schema_registry_ is `avro` but specified format in `WITH` clause is `protobuf`, an exception will be thrown.
+  * Schema with specified ID MUST exist in _schema_registry_, otherwise an exception will be thrown.
+  * Serde Features must be`WRAPPED` so that field names always exist. The reason is that if Serde Features has `UNWRAP_SINGLES` enabled, during translation, anonymous `ROWKEY` or `ROWVALUE` column name will be inserted for the schema first which makes the translated schema diverging from original schema in _Schema Registry_.
+  * Compatibility Checks. See section below.
+
+* `C*AS` command
+  * `key_format` and `value_format` are optional. If specified, they must match format fetched from _schema_registry_.
+  * Schema with specified ID MUST exist in _schema_registry_, otherwise an exception will be thrown.
+  * Serde Features must be `WRAPPED` so that field names always exist.
+  * Compatibility Checks. See section below.
+
+### Compatibility checks
+* Users can use `*_schema_id` with table elements in `CS/CT` command. For example, `CREATE STREAM stream_name (id INT KEY, value VARCHAR) with (kafka_topic='topic_name, value_schema_id=1, value_format='avro');`. In this case, schema fetched from _schema registry_ should be a superset of table elements and schema from _schema registry_ will be used as new schema.
+* Users can use `*_schema_id` with select projection in `C*AS` command. For example, `CREATE STREAM stream_name WITH (kafka_topic='topic_name', key_schema_id=1, value_schema_id=2,
+  partitions=1, key_format='avro', value_format='avro') AS SELECT * FROM source_stream;`. In this situation, schema from _schema registry_ should a superset of select projection and the extra fields should be optional fields or required fields with default value. This can make sure the new schema can be compatible with select projection schema which may have fewer fields.
```

Comment:
You're saying that ksql will only fetch the schema from Schema Registry at the time that the statement is submitted, after which ksql will write the fetched schema into the command itself (to be written into the command topic), which means ksql won't need to fetch the schema from Schema Registry while replaying from the command topic since the schema will already be embedded in the command itself?

If that's the case, then why are we writing `KEY_SCHEMA_ID` and `VALUE_SCHEMA_ID` into the plan at all?

Reply:
> You're saying that ksql will only fetch the schema from Schema Registry at the time that the statement is submitted, after which ksql will write the fetched schema into the command itself (to be written into the command topic), which means ksql won't need to fetch the schema from Schema Registry while replaying from the command topic since the schema will already be embedded in the command itself?

Yes.

> If that's the case, then why are we writing KEY_SCHEMA_ID and VALUE_SCHEMA_ID into the plan at all?
Which plan do you refer to? Query plan or plan in command written to command topic or QTT?

Reply:
I mean the plan that gets written into the command topic. That's the same as the plan.json file in QTT.

Reply:
The benefit would be make it clear the table elements is from schema registry. Maybe helping debugging?. I didn't realize it's in plan until I added it to format properties and saw test failure :) Is there downside putting it to plan? Storage increase?

Reply:
I thought we wrote the entire statement (including WITH clause properties) into the command topic as a text string. You should double-check me but if my understanding is correct, then for debugging purposes we're already covered.

The downside to writing it into the plan is (1) bloating the plan with unnecessary fields, and (2) added complexity to plan compatibility. You saw in your PR for writing schema ID into the plan that you had to generate new versions of the historic plans for the affected QTTs. We would not have to do that if we didn't write the unnecessary field into the plan.

Reply:
Synced offline. I will refactor the API to provide flag to control field name translation and remove the QTT plans generated with schema id